cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricStore.java

```diff
@@ -249,10 +267,27 @@ public void add(Collection<? extends MetricValues> metricValues) throws Exceptio
   }
 
   private CubeQuery buildCubeQuery(MetricDataQuery q) {
-    return new CubeQuery(null, q.getStartTs(), q.getEndTs(), q.getResolution(), q.getLimit(), q.getMetrics(),
+    String aggregation = getAggregation(q);
+    return new CubeQuery(aggregation, q.getStartTs(), q.getEndTs(), q.getResolution(), q.getLimit(), q.getMetrics(),
                          q.getSliceByTags(), q.getGroupByTags(), q.getInterpolator());
   }
 
+  @Nullable
+  private String getAggregation(MetricDataQuery q) {
+    // We mostly rely on auto-selection of aggregation during query (in which case null is returned from
+    // this method). In some specific cases we need to help resolve the aggregation though.
+    Set<String> tagNames = ImmutableSet.<String>builder()
+      .addAll(q.getSliceByTags().keySet()).addAll(q.getGroupByTags()).build();
+    if (tagNames.contains(Constants.Metrics.Tag.FLOW)) {
+      if (tagNames.contains(Constants.Metrics.Tag.PRODUCER) || tagNames.contains(Constants.Metrics.Tag.CONSUMER)) {
```

Comment:
just add a note here, why when we add user metrics for flow, we don't aggregate that in this aggregation group. since the size of this BY_FLOWLET_QUQUE is smaller compared to BY_FLOW, we need to choose appropriately, smth like that.